### PR TITLE
Implement letterspacing and under/overlines

### PR DESCRIPTION
**Pull request by rlafuente**
Picking up from #350

This is a working version of text() supporting `letterspacing`, `underline`, `underlinecolor`, `overline` and `overlinecolor`

Missing:
- also have them work in font()?
- color values have to be hex (or color name) to feed Pango, we can put in support for Color objects
- document, but would like to get the code+API right before that

### CODE REVIEW

File: shoebot/data/typography.py [hunk 1]
```diff
@@ -112,7 +112,8 @@ class Text(Grob, ColorMixin):
         Implementation of fonts uses Pango instead of Cocoa.
     """
 
-    # several reference docs can be found at http://www.pyGtk.org/docs/pygtk/class-pangofontdescription.html
+    # several reference docs can be found at
+    # https://developer.gnome.org/pygtk/stable/class-pangofontdescription.html
```
Comment (stuaxo) [inline shoebot/data/typography.py:116]: Is this right ?  Pygtk was the Gtk2 thing.

File: shoebot/data/typography.py [hunk 2]
```diff
@@ -178,7 +194,23 @@ def _pre_render(self):
...
+
+            # we want to output something like
+            # <span letter_spacing="2048">Hello World</span>
+            markup = '<span'
+            # markup_vars was set in __init__()
+            for setting, value in self.markup_vars.items():
+                markup += f' {setting}="{self.markup_vars[setting]}"'
+            markup += f">{self.text}</span>"
+            self._pango_layout.set_markup(markup)
+
```
Comment (stuaxo) [inline shoebot/data/typography.py:213]: I need to find some time to stare at this, it's def the sort of thing where tests are needed (think of how much trouble we got into with the nodebox colour lib)

Comment (stuaxo) [reply shoebot/data/typography.py:213]: Can you reply with some examples of the expected inputs (parameters to the method) and what the markup they should output ?

We can use this for writing some tests + find if there are any edge cases.

Comment (rlafuente) [reply shoebot/data/typography.py:213]: This is the format:
```
Inconsolata @wght=200
Inconsolata @wght=200,wdth=100
```

Comment (stuaxo) [reply shoebot/data/typography.py:213]: Cheers, so it would be:
```
 <span "Inconsolata @wght=200">blah blah</span>
 <span "Inconsolata @wght=200,wdth=100">blah blah</span>  
```

File: shoebot/data/typography.py (hunk 1 above)
Comment (rlafuente) [reply shoebot/data/typography.py:116]: Ah you're right! Still the API is pretty much the same, but in a quick search i couldn't find the current pango docs

Comment (rlafuente) [reply shoebot/data/typography.py:116]: I'll just remove the comment

File: shoebot/data/typography.py (hunk 2 above)
Comment (rlafuente) [reply shoebot/data/typography.py:213]: ah yes, exactly that! I could have typed out the whole thing, sorry

File: shoebot/data/typography.py (hunk 1 above)
Comment (stuaxo) [reply shoebot/data/typography.py:116]: Yeah I couldn't find the current docs either, I guess either remove it or mention it.

File: shoebot/data/typography.py (hunk 2 above)
Comment (stuaxo) [reply shoebot/data/typography.py:213]: np, unit tests are the definition of pedantry so it's good to capture all the details.

File: examples/demos/random-specimen.bot [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  # font title
+  fill(0.5)
+  font("Liberation Sans", 14)
+  text(randomfont, 15, y+30)
+
+  # specimen line
+  fill(0.2)
+  font(randomfont, 36)
+  text("Handgloves", 25, y+65)
+
```
Comment (stuaxo) [inline examples/demos/random-specimen.bot:13]: Def run Black on this.

File: shoebot/data/typography.py [hunk 4]
```diff
@@ -141,8 +142,23 @@ def __init__(
         self.font = kwargs.get("font", canvas.fontfile)
         self.fontsize = kwargs.get("fontsize", canvas.fontsize)
         self.align = kwargs.get("align", canvas.align)
-        self.indent = kwargs.get("indent")
         self.lineheight = kwargs.get("lineheight", canvas.lineheight)
+        self.indent = kwargs.get("indent")
```
Comment (stuaxo) [inline shoebot/data/typography.py:146]: I guess we can't set indent on canvas ?

File: shoebot/data/typography.py [hunk 5]
```diff
@@ -178,7 +194,23 @@ def _pre_render(self):
         # self.layout.set_spacing(int(((self.lineheight-1)*self._fontsize)*Pango.SCALE)) #pango requires an int casting
         # we pass pango font description and the text to the pango layout
         self._pango_layout.set_font_description(self._pango_fontface)
-        self._pango_layout.set_text(self.text, -1)
+
+        if not self.markup_vars:
+            self._pango_layout.set_text(self.text, -1)
```
Comment (stuaxo) [inline shoebot/data/typography.py:199]: This looks like an early bailout, which is good - its simplifies things.
Put a `return` here and ditch the `else` that follows and all the code below can lose an indent level.

Comment (stuaxo) [reply shoebot/data/typography.py:199]: This `if` is also the boundry between two tests:  in theory at least, every `if` generates a new test.   Which might be why one reason test people try and avoid state where possible :)

The rest of this will be about avoiding state.

Where possible, state can be avoided by avoiding:
- `.append()` on `list`s, by using list comprehensions and filters.*
- avoiding `+=` to build strings using fstrings, and `.join` on strings.*
- avoiding setting elements in dicts*

* within reason !  these aren't hard and fast rules.

Comment (stuaxo) [reply shoebot/data/typography.py:199]: I'll document attempting to refactor this so there's less code to test:

Maybe this can be refactored and use dict comprehensions / filters to reduce the manual dict building.
There are two things going on in this loop, so it makes sense to separate them:

1.  Copying named variables `("underline  ...`
2. Creating the `letter_spacing` variable from `tracking`

For 1:
It's a visit to the trusty search engine with the search terms:   `python dict filter certain items` and luckily there's this answer https://stackoverflow.com/questions/23862406/filter-items-in-a-python-dictionary-where-keys-contain-a-specific-string

In the end I couldn't see anything exactly right, so tried a couple of things in jupyter using the expected inputs + outputs and got it down to this:
```markup_vars = {arg: kwargs[arg] for arg in ["underline", "underlinecolor"] if arg in kwargs}```
(I was only testing with the first couple of arguments)

For 2:
To keep this simple it may as well adjust the dict, but is moved outside the loop:
```
markup_vars = {arg: kwargs[arg] for arg in ["underline", "underlinecolor"] if arg in kwargs}
if "tracking" in kwargs:
    markup_vars["letter_spacing"] = int(1024 * kwargs["tracking"])
```

(I haven't run this through Black, or added all the needed variables).

File: shoebot/data/typography.py [hunk 6]
```diff
@@ -178,7 +194,23 @@ def _pre_render(self):
...
+            # and https://developer.gnome.org/pango/1.46/pango-Markup.html
+
+            # we want to output something like
+            # <span letter_spacing="2048">Hello World</span>
+            markup = '<span'
+            # markup_vars was set in __init__()
+            for setting, value in self.markup_vars.items():
+                markup += f' {setting}="{self.markup_vars[setting]}"'
+            markup += f">{self.text}</span>"
+            self._pango_layout.set_markup(markup)
```
Comment (stuaxo) [inline shoebot/data/typography.py:212]: It looks like the same kind of refactors may be possible - I'll document the process:

Ideas before starting:
- Having <span> .. </span> split at the beginning and end, while logical can be hard to follow, maybe it'll be possible to do something like f"<span {markup}>{some_data}</span>

OK, lets go...

The first thing, is this, code:
```python
for setting, value in self.markup_vars.items():
    markup += f' {setting}="{self.markup_vars[setting]}"'
```

Can be simplified by using the `value` variable it already has:
```python
for setting, value in self.markup_vars.items():
    markup += f' {setting}="{value}"'
```

Then after a little prototyping in jupyter it became obvious, this could be changed to a list comprehension, suitable for passing to `" ".join(... )` later:
```python
[f'{setting}={value}' for setting, value in self.markup_vars.items()]
['underline=1', 'underlinecolor=#eeaa00']
```

Next step is to convert it to use str.join:
```python
" ".join([f'{setting}={value}' for setting, value in self.markup_vars.items()])
'underline=1 underlinecolor=#eeaa00'
```

And then <span> can all be done in one place and hopefully the eye doesn't have to jump around as much to follow the logic:
```python
markup_styles = " ".join([f'{setting}="{value}"' for setting, value in self.markup_vars.items()])
self._pango_layout.set_markup(f"<span {markup_styles}>{self.text}</span>")
```

(not run through Black yet).

Comment (stuaxo) [reply shoebot/data/typography.py:212]: I did more of this than I would normally to give some visibility into process, but hopefully move things to the test side sooner.
Next time I'll try and get a Mock going with Pango, then you can call text() from your test and verify it is called with the expected arguments + we should have a nice e2e test going.

File: shoebot/data/typography.py (hunk 4 above)
Comment (rlafuente) [reply shoebot/data/typography.py:146]: I'm not sure why, maybe it sat there unfinished? I can try putting it back later.

File: shoebot/data/typography.py (hunk 6 above)
Comment (rlafuente) [reply shoebot/data/typography.py]: CHEERS
all done!

File: shoebot/data/typography.py (hunk 2 above)
Comment (rlafuente) [reply shoebot/data/typography.py]: wait, according to the [current docs](https://lazka.github.io/pgi-docs/Pango-1.0/), the syntax should be:
```
"Cantarell Italic Light 15 `wght`=200"
```
we should update this (i can do it, just not now)

File: examples/demos/random-specimen.bot [hunk 7]
```diff
@@ -0,0 +1,12 @@
+for x,y in grid(1, 6, 10, 60):
```
Comment (stuaxo) [inline examples/demos/random-specimen.bot:1]: formatting: `x, y`

File: shoebot/data/typography.py (hunk 4 above)
Comment (rlafuente) [reply shoebot/data/typography.py:146]: I think it can be set as a kwarg to text(), but it is not part of the state variables like fontsize (which have functions to set them)

File: shoebot/data/typography.py (hunk 2 above)
Comment (rlafuente) [reply shoebot/data/typography.py]: OK so I tested this syntax and it doesn't seem to work. I'll stick to the @-syntax and write a small test to ensure markup is correct

Comment (rlafuente) [reply shoebot/data/typography.py]: I see no way to get the markup in Pango.Layout (only set), and we're not holding the markup data anywhere, so I'm a bit stumped as how to test this.

Comment (rlafuente) [reply shoebot/data/typography.py]: OR, for a minimal test, we can just query font():

```python
font("Inconsolata", var_wdth=100, var_wght=200*
print(font())
```

gives

```
Inconsolata @wdth=100,wght=200
```

Comment (stuaxo) [reply shoebot/data/typography.py]: That sounds like a sensible test.

I already added `test_text_outputs_pango_text_spans` on this branch, which makes an effort, using mocks to see if we pass through what's expected to Pango.

Comment (rlafuente) [reply shoebot/data/typography.py]: done! making tests is a joy now, this framework is just great. Cheers :)

**Review by stuaxo — APPROVED** [review]
